## Тема 2.1 Введение в ML. Основные задачи, виды классических моделей, метрики качества моделей.

### 1. Обучение первой модели

_Характеристики набора данных_:

- **Количество экземпляров**: 150 (по 50 в каждом из трех классов)

- **Количество атрибутов**: 4 числовых, прогнозных атрибута и класс

- **Информация об атрибутах**:

    - длина чашелистика (sepal) в см
    - ширина чашелистика (petal) в см
    - длина лепестка в см
    - ширина лепестка в см
    - класс:
        - Ирис-Сетоса (setosa)
        - Ирис-разноцветный (versicolor)
        - Ирис-Вирджиния (virginica)

https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_iris.html

Задача:

Построить модель машинного обучения для решения задачи регрессии: необходимо спрогнозировать прогресс заболевания диабетом через год после исходного уровня.

План решения:
1. В документации библиотеки scikit learn найдите, как загрузить датасет для построения модели прогноза прогрессирования заболевания через год после исходного уровня (load_diabetes из sklearn.datasets).

In [41]:
from sklearn.datasets import load_diabetes
from sklearn import tree
from sklearn.metrics import mean_squared_error, root_mean_squared_error, r2_score, mean_absolute_error
from sklearn.model_selection import train_test_split
from matplotlib import pyplot as plt
from matplotlib import pyplot as plt
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression

In [42]:
df = load_diabetes(as_frame=True).frame
df.head().T

0          1           2           3           4
age       0.038076  -0.001882    0.085299   -0.089063    0.005383
sex       0.050680  -0.044642    0.050680   -0.044642   -0.044642
bmi       0.061696  -0.051474    0.044451   -0.011595   -0.036385
bp        0.021872  -0.026328   -0.005670   -0.036656    0.021872
s1       -0.044223  -0.008449   -0.045599    0.012191    0.003935
s2       -0.034821  -0.019163   -0.034194    0.024991    0.015596
s3       -0.043401   0.074412   -0.032356   -0.036038    0.008142
s4       -0.002592  -0.039493   -0.002592    0.034309   -0.002592
s5        0.019907  -0.068332    0.002861    0.022688   -0.031988
s6       -0.017646  -0.092204   -0.025930   -0.009362   -0.046641
target  151.000000  75.000000  141.000000  206.000000  135.000000

2. Выведите содержимое поля DESCR, которое вернет функция, загружающая датасет, чтобы изучить содержимое датасета.



In [43]:
print(load_diabetes().DESCR)

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

:Number of Instances: 442

:Number of Attributes: First 10 columns are numeric predictive values

:Target: Column 11 is a quantitative measure of disease progression one year after baseline

:Attribute Information:
    - age     age in years
    - sex
    - bmi     body mass index
    - bp      average blood pressure
    - s1      tc, total serum cholesterol
    - s2      ldl, low-density lipoproteins
    - s3      hdl, high-density lipoproteins
    - s4      tch, total cholesterol / HDL
    - s5      ltg, possibly log of serum triglycerides level
    - s6      glu, blood sugar level

Note: Each of these 10 feature variables have bee

In [44]:
df.describe().T

count          mean        std        min        25%         50%  \
age     442.0 -2.511817e-19   0.047619  -0.107226  -0.037299    0.005383   
sex     442.0  1.230790e-17   0.047619  -0.044642  -0.044642   -0.044642   
bmi     442.0 -2.245564e-16   0.047619  -0.090275  -0.034229   -0.007284   
bp      442.0 -4.797570e-17   0.047619  -0.112399  -0.036656   -0.005670   
s1      442.0 -1.381499e-17   0.047619  -0.126781  -0.034248   -0.004321   
s2      442.0  3.918434e-17   0.047619  -0.115613  -0.030358   -0.003819   
s3      442.0 -5.777179e-18   0.047619  -0.102307  -0.035117   -0.006584   
s4      442.0 -9.042540e-18   0.047619  -0.076395  -0.039493   -0.002592   
s5      442.0  9.293722e-17   0.047619  -0.126097  -0.033246   -0.001947   
s6      442.0  1.130318e-17   0.047619  -0.137767  -0.033179   -0.001078   
target  442.0  1.521335e+02  77.093005  25.000000  87.000000  140.500000   

               75%         max  
age       0.038076    0.110727  
sex       0.050680    0.050680  
bmi       0.031248    0.170555  
bp        0.035644    0.132044  
s1        0.028358    0.153914  
s2        0.029844    0.198788  
s3        0.029312    0.181179  
s4        0.034309    0.185234  
s5        0.032432    0.133597  
s6        0.027917    0.135612  
target  211.500000  346.000000

In [45]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 442 entries, 0 to 441
Data columns (total 11 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   age     442 non-null    float64
 1   sex     442 non-null    float64
 2   bmi     442 non-null    float64
 3   bp      442 non-null    float64
 4   s1      442 non-null    float64
 5   s2      442 non-null    float64
 6   s3      442 non-null    float64
 7   s4      442 non-null    float64
 8   s5      442 non-null    float64
 9   s6      442 non-null    float64
 10  target  442 non-null    float64
dtypes: float64(11)
memory usage: 38.1 KB


3. Подготовьте данные к обучению моделей: отделите целевой признак от датасета, разделите датасет на тренировочную и валидационную части.



In [46]:
df['target'].value_counts()

target
200.0    6
72.0     6
90.0     5
178.0    5
71.0     5
        ..
73.0     1
222.0    1
86.0     1
79.0     1
57.0     1
Name: count, Length: 214, dtype: int64

In [47]:
features = df.drop(['target'], axis=1)
target = df['target']


In [48]:
features_train, features_val, target_train, target_val = train_test_split(features, target, test_size=0.2, random_state=17)

4. Обучите решающее дерево и оцените адекватность обученной модели. Для этого достаточно сравнить метрики качества с метриками, получаемыми от простой модели. Например, если бы мы подавали на вход среднее значение целевой переменной на обучающей части выборки.

In [49]:
model = tree.DecisionTreeRegressor()
model.fit(features_train, target_train)
y_pred = model.predict(features_val)

def count_errors(true_answers, pred_answers):
    all_answers_together = zip(true_answers, pred_answers)
    errors_list = [1 if v[0] != v[1] else 0 for v in all_answers_together]
    return sum(errors_list)

print("Ошибок:", count_errors(target_val, y_pred))

mse = mean_squared_error(target_val, y_pred)
rmse= np.sqrt(mse)
r2 = r2_score(target_val, y_pred)
mae = mean_absolute_error(target_val, y_pred)

print("nОбученная модель:")
print("MSE: ", mse)
print("RMSE: ", rmse)
print("R2: ", r2)
print("MAE: ", mae)

Ошибок: 87
nОбученная модель:
MSE:  6003.9775280898875
RMSE:  77.48533750387803
R2:  -0.016722977032058273
MAE:  61.59550561797753


Чтобы оценить адекватность модели, создаю простую модель со средним значением целевой переменной на входе.

In [50]:
target_avg_val = pd.Series([target.mean()] * len(target_val))

In [51]:
print("Ошибок:", count_errors(target_val, target_avg_val))

mse_simple = mean_squared_error(target_val, target_avg_val)
rmse_simple= np.sqrt(mse_simple)
r2_simple = r2_score(target_val, target_avg_val)
mae_simple = mean_absolute_error(target_val, target_avg_val)

print("простая модель:")
print("MSE: ", mse_simple)
print("RMSE: ", rmse_simple)
print("R2: ", r2_simple)
print("MAE: ", mae_simple)

Ошибок: 89
простая модель:
MSE:  5928.798335587457
RMSE:  76.9986904796923
R2:  -0.0039920145901892745
MAE:  67.71236463470436


Вывод: 

Исходя из полученных результатов делаем вывод, что обученная модель адекватна.

5. Обучите линейную регрессию и оцените ее адекватность.

In [52]:
model_regr = LinearRegression()
model_regr.fit(features_train, target_train)
y_pred_regr = model_regr.predict(features_val)

print("Ошибок:", count_errors(target_val, y_pred_regr))

mse_regr = mean_squared_error(target_val, y_pred_regr)
rmse_regr= np.sqrt(mse_regr)
r2_regr = r2_score(target_val, y_pred_regr)
mae_regr = mean_absolute_error(target_val, y_pred_regr)

print("nОбученная модель:")
print("MSE: ", mse_regr)
print("RMSE: ", rmse_regr)
print("R2: ", r2_regr)
print("MAE: ", mae_regr)

Ошибок: 89
nОбученная модель:
MSE:  3231.41947978159
RMSE:  56.84557572741778
R2:  0.4527863540208549
MAE:  45.89989611995773


оцениваем на адекватность

In [53]:
print("Ошибок:", count_errors(target_val, target_avg_val))

mse_regr_simple = mean_squared_error(target_val, target_avg_val)
rmse_regr_simple= np.sqrt(mse_regr_simple)
r2_regr_simple = r2_score(target_val, target_avg_val)
mae_regr_simple = mean_absolute_error(target_val, target_avg_val)

print("nОбученная модель:")
print("MSE: ", mse_regr_simple)
print("RMSE: ", rmse_regr_simple)
print("R2: ", r2_regr_simple)
print("MAE: ", mae_regr_simple)

Ошибок: 89
nОбученная модель:
MSE:  5928.798335587457
RMSE:  76.9986904796923
R2:  -0.0039920145901892745
MAE:  67.71236463470436


Вывод:

Исходя из полученных результатов делаем вывод, что модель адекватна.

6. Выберите лучшую модель и обоснуйте свой выбор.



Показатели для решающего дерева:

Ошибок: 87

MSE:  6315.044943820225

RMSE:  79.4672570548413

R2:  -0.06939962138979205

MAE:  62.97752808988764

показатели для линейной регрессии:

Ошибок: 89

MSE:  3231.41947978159

RMSE:  56.84557572741778

R2:  0.4527863540208549

MAE:  45.89989611995773

Сравнив полученые метрики делаем вывод, что линейная регрессия лучше